internal/secretsource/url/url.go

```diff
@@ -0,0 +1,401 @@
+			if isRetryableError(response.StatusCode, nil) {
+				return &retryableError{err: statusErr, statusCode: response.StatusCode}
+			}
+
+			// Non-retryable error (4xx except 429)
+			return statusErr
+		}
+
+		// Extract secret value from response
+		extractedSecret, err := extractSecretFromResponse(response.Body, us.config.ResponsePath)
+		if err != nil {
+			// Extraction errors are not retryable (indicates config issue)
+			return fmt.Errorf("failed to extract secret: %w", err)
+		}
+
+		secret = extractedSecret
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return secret, nil
+}
+
+type limiter interface {
+	Wait(ctx context.Context) error
+}
+
+func newLimiter(requestsPerMinuteLimit, requestsBurst int) *rate.Limiter {
+	// The calculation below looks wrong because it seems like it's giving
+	// n min²/req, but the first number is actually time unit/min, so the
+	// units of the result are time unit/req, which is correct because it's
+	// the interval of time after which a new token is replenished. In
+	// other words, the units are time unit/token.
+	tokenReplenishInterval := time.Minute / time.Duration(requestsPerMinuteLimit)
+
+	return rate.NewLimiter(rate.Every(tokenReplenishInterval), requestsBurst)
+}
+
+func extractSecretFromResponse(body io.Reader, responsePath string) (string, error) {
+	data, err := io.ReadAll(body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	if responsePath == "" {
+		// If no path specified, assume response body is the secret
+		return string(data), nil
+	}
+
+	result := gjson.GetBytes(data, responsePath)
+
+	if !result.Exists() {
+		return "", fmt.Errorf("path %q not found in response", responsePath)
+	}
+
+	if result.Type != gjson.String {
+		return "", fmt.Errorf("secret value at path %q is not a string (got %s)", responsePath, result.Type)
+	}
+
+	return result.String(), nil
+}
+
+// isRetryableError determines if an HTTP status code or error should trigger a retry.
+// Retries on:
+// - Network errors (connection failures, timeouts)
+// - 5xx server errors (server-side issues)
+// - 429 Too Many Requests (rate limiting)
+// Does NOT retry on:
+// - 4xx client errors (except 429) - these indicate issues with the request itself
+func isRetryableError(statusCode int, err error) bool {
+	// Network errors should be retried
+	if err != nil {
+		return true
+	}
+
+	// Retry on server errors and rate limiting
+	if statusCode >= 500 || statusCode == http.StatusTooManyRequests {
+		return true
+	}
+
+	return false
+}
+
+// retry retries to execute a provided function until it succeeds or the maximum
+// number of attempts is hit. It waits with exponential backoff between attempts.
+// The backoff calculation is: wait = (base ^ attempt) + random jitter up to 1 second.
+// Only retries errors that are of type *retryableError.
+func retry(ctx context.Context, attempts int, baseBackoff time.Duration, do func() error) error {
```

Comment:
I feel like it would've been easier if this is done with a function returning directly a boolena for whether it is retryable instead of wrapping the error. I am pretty sure you will cut 10-20 lines of code as well.